Added verbose option that returns detailed output from tesseract run

REMitchell opened the pull request:
I had a project that required getting confidence values for the Tesseract run and saw that this library didn't support tsv files, so here's a proposal for a feature that I think will add a lot of value for users. The tsv file contains information about box positions/sizes, line and page numbers, and confidences, as well as providing the text. 

While this update to pytesseract is compatible with Python 2.4+ the tsv feature only exists in tesseract 3.05+

Current output:

```
level	page_num	block_num	par_num	line_num	word_num	left	top	width	height	conf	text
1	1	0	0	0	0	0	0	600	103	-1	
2	1	1	0	0	0	23	26	555	51	-1	
3	1	1	1	0	0	23	26	555	51	-1	
4	1	1	1	1	0	23	26	555	22	-1	
5	1	1	1	1	1	23	26	44	17	86	This
5	1	1	1	1	2	76	26	15	17	86	is
5	1	1	1	1	3	99	30	57	13	86	some
5	1	1	1	1	4	164	26	42	21	73	text,
5	1	1	1	1	5	215	26	70	17	82	written
5	1	1	1	1	6	295	26	15	17	95	in
5	1	1	1	1	7	317	26	52	21	73	Arial,
5	1	1	1	1	8	378	26	40	17	80	that
5	1	1	1	1	9	425	26	32	17	95	will
5	1	1	1	1	10	467	26	24	17	85	be
5	1	1	1	1	11	500	26	45	17	87	read
5	1	1	1	1	12	555	26	23	22	85	by
4	1	1	1	2	0	23	54	513	23	-1	
5	1	1	1	2	1	23	55	107	17	82	Tesseract.
5	1	1	1	2	2	140	55	49	17	91	Here
5	1	1	1	2	3	198	59	32	13	86	are
5	1	1	1	2	4	239	59	57	13	86	some
5	1	1	1	2	5	304	55	92	22	86	symbols:
5	1	1	1	2	6	407	54	129	23	65	!@#$%"&‘()
```

int3l commented:
Hello @REMitchell , Thank you very much for your contribution!
We already have ongoing discussion about the different outputs and output formats.
Currently it's a limitation for the pytesseract package.

The only downside for your addition is that the name of the function (image_to_string) suggest string result and it will be a little bit confusing for the end user. Keep in mind that there is no proper documentation for pytesseract, except the README file.

Anyway, that doesn't mean that your contribution is not accepted - we just may need to extract the current logic for it, in a separate function.

For example it can be **image_to_tsv** or as we discussed in #75 , we can have generic function **image_to_file**, which can allow more generic approach for handling the different output formats of tesseract.

Also beforehand, it will be a good idea to extract the common logic from image_to_string function, so it can be reused (it will prevent DRY and will allow easier maintaining/support).

REMitchell commented:
@int3l Great idea! Completely overlooked the **image_to_string** naming problem. 
This sounds like you're proposing a major change, but if you don't mind the upheaval, I certainly don't mind writing the code!
Will work on this tonight.

int3l commented:
It's a big structural change for sure :) But this is the cost for making the pytesseract package more useful and with more features.

If we do it carefully/correctly, it will be transparent for the end user and it shouldn't cause (noticeable) upheaval :D
So go for it, if you want and have the time.

PS: It is on my TODO list too, but I didn't have the chance to start working on it yet.

int3l commented on src/pytesseract.py:
```diff
@@ -70,6 +71,9 @@ def run_tesseract(input_filename,
     command += (tesseract_cmd, input_filename, output_filename_base)
 
+    if verbose:
+        command.append('tsv')
```
Just a note about the output configs:
The official documentation states that the output config option is appended at the end of the cli command.
I am not sure if this is really the case, but I guess that for the sake of consistency with the official docs, maybe we can append the output config options at the end of the list with arguments?
(Also, in this way we may avoid any argument parsing bugs during the tesseract invocation)

REMitchell commented:
@int3l Done. 

I realized that, going the route that we are, it might also be appropriate to extract out a "image_to_boxes" function, so I did that. Added backwards compatibility and a warning print statement in the meantime.

A couple improvements that could be made (potentially in a separate PR):

- You have a lot of information in the Docstrings and, with three similar functions (and very similar, Docstrings) they're getting a little wordy. Should I make them a little more concise, move things over the README, and tell users to RTM?

- Return data as tsv seems a little unpythonic and, if users are really going to be using this in code, they're probably going to need to parse this output into some more usable object anyway. I know I started to do that in the previous incarnation of this PR, but do you think it would make sense to do something a little nicer for both box and tsv files?

REMitchell commented on src/pytesseract.py:
```diff
@@ -94,55 +102,54 @@ def prepare(image):
     raise TypeError('Unsupported image object')
 
+def run_and_get_output(image, extension, lang=None, config=None, nice=None):
+    temp_name = ''
```
Avoids additional exceptions in the stack trace if an exception occurs in save_image

int3l commented:
First of all, thank you for the contribution again!

About the first point:
My opinion is the same. The docstrings are too verbose, especially after the refactoring.

About the second one:
I agree, the parsing implementation is already available thanks to you, so it can be preserved.

As far as about the way we enable the parsing - it can be an additional parameter.
Or if we don't want to overload the functions with additional parameters - it can be just an additional separate function that takes as input the result from our newly available **image_to_*** functions.

But in any case it still will be a good idea to keep the tsv/etc. parsing logic (and in an encapsulated way as of a function), so the end user can have more options like "custom parsing combined with the standard parsing function" or etc.

int3l commented on src/pytesseract.py:
```diff
@@ -98,63 +102,54 @@ def prepare(image):
+        cleanup(temp_name)
 
-def image_to_string(image, lang=None, boxes=False, config=None, nice=0, verbose=0):
+def image_to_string(image, lang=None, config=None, nice=0):
```
One small nitpick - since the image_to_string function signature is changed, some end users might get TypeError for unexpected argument, because the boxes flag is not available anymore.

Maybe we can add deprecation warning that suggests using the newly available function and aborting the function gracefully, so the users can migrate more easily, instead of just getting confusing exceptions at first.

I can include this if you want.

REMitchell replied:
I actually had this change locally and forgot to commit it. My thoughts exactly :)

int3l commented:
Hi, sorry for the delay, but I'm a bit ill.
If you don't have any other changes, please let me know, so I can accept the PR.

anupamme commented:
Hello,

I need to use the changes made in this issue/pull request. Which branch should I be using?

REMitchell commented:
@anupamme Glad to hear it! You can use this repo if you need to, but hopefully these changes should get merged in soon: https://github.com/REMitchell/pytesseract

REMitchell commented:
@int3l No problem. Just did some cleanup things, added to the README (you might want to give that a lookover -- I'm more familiar with the .md format...) 

Added a simple string file to dict parser function, with a boolean ```dict_output``` optional parameter to the three user functions.

Also cleaned up the docstrings!

int3l commented:
Merging the current changes to master for further testing and preparation for release cut.
